### PR TITLE
Keycloak Backend Functionality

### PR DESCRIPTION
**Pull request by ZachBourqueBCGov**
# Summary

Overhauled the backend keycloak implementation to better match [this example app](https://github.com/bcgov/keycloak-example-apps/tree/dev/education/oidc/confidential/nextjs/backend).

This PR only provides backend functionality. Frontend storage of the access token and usage of said access token is still to come.

# How To Test

### Login/Logout Functionality
1. Replace the "login" button from lines 50-55 of the frontend file "LandingPage.jsx" with the following:
```jsx
  <a
    className="btn btn__primary"
    style={{ maxWidth: "150px" }}
    href="http://localhost:5005/oauth/login"
  >
    Login with IDIR
  </a>
  <a
    className="btn btn__primary"
    style={{ maxWidth: "150px" }}
    href="http://localhost:5005/oauth/logout"
  >
    Logout
  </a>
```
2. Login by clicking the `Login with IDIR` button
3. Ensure that you are logged in by viewing the `token` field in the URL
image
4. Check the `Application` tab in your browser's devtools to ensure the refresh_token httpOnly cookie is set
5. Click the logout button and ensure that both the token field in the URL and the refresh_token cookie are unset

### Endpoint protection functionality

1. Login to the api using the frontend
2. Copy the token query param from the URL
3. Using a request client of your choice, make a request to the `/users` API endpoint, with the given token in the Authorization header
  a. Ensure you get a successful response
  b. Remove the token and ensure an unsuccessful response

[](https://hamzamohdzubair.github.io/redant/)

- None listed.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue).
- [ ] Dependency updates (may include refactoring of a dependencies functionality).
- [x] New feature (non-breaking change which adds functionality).

## Checklist:

- [ ] I have read the checklist.

### CODE REVIEW

**Comment by Bottle7** [comment]
This takes me to a 'Connection was reset..." page. For logging in and logging out links.

**Review by Bottle7 — APPROVED** [review]
🎸

File: matchmaking/backend/config.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+require("dotenv").config({ path: __dirname + "/../../.env" });
```
Comment (BradyMitch) [inline matchmaking/backend/config.js:1]: You shouldn't need this as dotenv is required in index.js which makes process.env available to the rest of the app. Also you can't set the dirname to ../../.env as that will only work locally. The its left as default in index.js and grabs the env vars set by the docker-compose and dockerfile

**Comment by BradyMitch** [comment]
How do we access the user info on a request now? Like user guid

File: matchmaking/backend/config.js (hunk 1 above)
Comment (ZachBourqueBCGov) [reply matchmaking/backend/config.js:1]: Good catch, I'll make the change

**Comment by ZachBourqueBCGov** [comment]
> How do we access the user info on a request now? Like user guid

Because we have access to the authorization header for each request, in the future we can add a middleware to decode and map the JWT in said header to somewhere on the request object. We could also modify the protect middleware to do the same thing, but I feel the functionality is different enough that they should be separate middleware.

**Pull request merged**